### PR TITLE
(bpranaw's code) Camera capture integration

### PR DESCRIPTION
**Pull request by nrwahl2**

### CODE REVIEW

File: oresat_dxwifi/camera/camera.py [hunk 1]
```diff
@@ -1,107 +1,234 @@
...
+        )
+
+    def check_seconds_per_video(self, duration, seconds_per_video):
+        is_valid = True
+        if duration % seconds_per_video != 0:
+            is_valid = False
+        return is_valid
+
+    # Make or overwrite the given directory
+    def fresh_dir(self, path):
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py]: Anything that we turn into a method, if it doesn't use its `self` argument it can be a `staticmethod`.

We can also prepend an underscore to the name of any helper method that users of the class won't call (e.g., `def fresh_dir -> def _fresh_dir`), to emphasize that they're intended to be private.

File: oresat_dxwifi/camera/camera.py [hunk 2]
```diff
@@ -1,107 +1,234 @@
...
+    )
+    parser.add_argument(
+        "-vx",
+        "--video-x",
+        type=int,
+        default=640,
+        help="Horizontal resolution of video.",
+    )
+    parser.add_argument(
+        "-vy", "--video-y", type=int, default=480, help="Vertical resolution of video."
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py]: If we're doing all the rest one-argument-per-line, then I'd do the same here for consistency even though it fits.

File: oresat_dxwifi/camera/camera.py [hunk 3]
```diff
@@ -1,107 +1,234 @@
...
-        print("Cleaned Directory: {}".format(path))
-    else:
-        # Make
-        os.mkdir(path)
-        print("Created new directory: {}".format(path))
-
-# Main script
-if __name__ == "__main__":
 
+def define_parse_arguments():
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py]: I'd call this `create_parser()` for clarity

File: oresat_dxwifi/camera/camera.py [hunk 4]
```diff
@@ -1,107 +1,234 @@
...
+        default=False,
+        action="store_true",
+        help="Delete the raw frames after completion.",
+    )
+
+    return parser
+
+
+class CameraV4l2Interface:
+    def __init__(
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py:14]: Type hints would be good here and elsewhere

File: oresat_dxwifi/camera/camera.py [hunk 5]
```diff
@@ -1,107 +1,234 @@
...
+                except OSError:
+                    os.remove(p)
+                    print("Removed file: {}".format(f))
+            print("Cleaned Directory: {}".format(path))
+        else:
+            # Make
+            os.mkdir(path)
+            print("Created new directory: {}".format(path))
+
+    def overwrite_directories(self, frames_directory_path, video_directory_path):
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py]: I don't think we need this method. Instead, `create_videos()` can call `self._fresh_dir()` twice, once with each directory as the argument.

I can't think of a good reason a user of the class would need to call `overwrite_directories()` directly.

File: oresat_dxwifi/camera/camera.py [hunk 6]
```diff
@@ -1,107 +1,234 @@
...
+        args.bit_rate,
+        args.total_duration,
+        args.image_output,
+        args.video_output,
+        args.delete_frames
+    )
+
     # Basic check -- make sure seconds per video divides total duration
-    if args.total_duration % args.spv != 0:
+    if not cam_interface.check_seconds_per_video(
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py]: Validations can be part of the `__init__()` method and can throw an exception. We can handle the exception in the caller or simply include enough info in the exception itself about the error.

File: oresat_dxwifi/camera/camera.py [hunk 7]
```diff
@@ -1,107 +1,234 @@
...
+
+    def overwrite_directories(self, frames_directory_path, video_directory_path):
+        # Make / overwrite directories
+        self.fresh_dir(frames_directory_path)
+        self.fresh_dir(video_directory_path)
+
+    def calculate_number_of_loops(self):
+        return self.total_duration // self.seconds_per_video
+
+    def delete_frames(self):
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py]: We need to call `delete_frames()`

File: oresat_dxwifi/camera/camera.py [hunk 8]
```diff
@@ -1,107 +1,234 @@
...
+            "-c:v libx264 -b:v {}k -preset ultrafast -loglevel quiet -an -y".format(
+                self.bit_rate
+            )
+        )
+
+    def check_seconds_per_video(self, duration, seconds_per_video):
+        is_valid = True
+        if duration % seconds_per_video != 0:
+            is_valid = False
+        return is_valid
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py:71]: one-liner: `return (duration % seconds_per_video) == 0`

File: oresat_dxwifi/camera/camera.py [hunk 9]
```diff
@@ -1,107 +1,234 @@
...
+            # Make
+            os.mkdir(path)
+            print("Created new directory: {}".format(path))
+
+    def overwrite_directories(self, frames_directory_path, video_directory_path):
+        # Make / overwrite directories
+        self.fresh_dir(frames_directory_path)
+        self.fresh_dir(video_directory_path)
+
+    def calculate_number_of_loops(self):
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py]: IMO, better to put `self.total_duration // self.seconds_per_video` directly in the loop header and drop this method. It's just one arithmetic operation and it's called only once. Making it a separate method obfuscates what's happening, without making things more readable.

File: oresat_dxwifi/camera/camera.py [hunk 10]
```diff
@@ -1,107 +1,234 @@
...
+        spv,
+        br,
+        duration,
+        image_output,
+        video_output,
+        frames_del,
+    ):
+        self.camera_binary_path = binary
+        self.device_path = device
+        self.horizontal_video_resolution = video_x
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py]: `self.horizontal_resolution` or `self.x_resolution` or `self.x_res` or `self.res_x`. This is used for the raw frames too (not just video), and "video" doesn't add info.

File: oresat_dxwifi/camera/camera.py [hunk 11]
```diff
@@ -1,107 +1,234 @@
...
+        self.vertical_video_resolution = video_y
+        self.frames_per_second = fps
+        self.seconds_per_video = spv
+        self.bit_rate = br
+        self.total_duration = duration
+        self.image_output_directory_path = image_output
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py]: `image_output_directory` or `image_output_path`

Likewise for video output below

File: .gitignore [hunk 12]
```diff
@@ -127,3 +127,22 @@ dmypy.json
 
 # Pyre type checker
 .pyre/
+
```
Comment (nrwahl2) [inline .gitignore:130]: To be clear: Is everything in this commit taken (without changes) from `oresat-live-software/camera`?

File: oresat_dxwifi/camera/new_capture_test/test_capture.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+from io import BytesIO
+from PIL import Image
+
+class CameraInterface:
+    def __init__(
+        self,
+        buffer_type=BufferType.VIDEO_CAPTURE,
+        horizontal_resolution=640,
+        vertical_resolution=480,
+        file_format="MJPG",
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/new_capture_test/test_capture.py:22]: Any particular reason for the choice of `"MJPG"`, or just a placeholder for now? I don't know format trade-offs, off the top of my head.

We need to at least check output video sizes. I think `ffmpeg` performs compression, while this may not (or this may not compress it as much).

File: oresat_dxwifi/camera/new_capture_test/test_capture.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+        self.fps = fps
+
+    def capture(self) -> None:
+        # Just to see whats going on
+        fmt = "%(threadName)-10s %(asctime)-15s %(levelname)-5s %(name)s: %(message)s"
+        logging.basicConfig(level="INFO", format=fmt)
+
+        # You'll have to check /dev/v4l/by-id/ to see what your video device is listed under.
+        # It changed on my BBB for seemingly no reason.
+        CAMERA_LOCATION = "/dev/v4l/by-id/usb-Empia._USB_Camera_SN202106-video-index0"
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/new_capture_test/test_capture.py:38]: Probably not something we want to hard-code anyway. Perhaps we could default it to a symlink that's more generic than this path that has the camera model (assuming one exists somewhere in `/dev`).

**Review by nrwahl2 — COMMENTED** [review]
Looking good so far. I understand it's a work in progress. Comments on the current state.

File: .gitignore (hunk 12 above)
Comment (bpranaw) [reply .gitignore:130]: I specified then directories oresat_dxwifi/camera but yeah these are all from the gitignore of oresat-live-software/camera

File: oresat_dxwifi/camera/new_capture_test/test_capture.py (hunk 13 above)
Comment (bpranaw) [reply oresat_dxwifi/camera/new_capture_test/test_capture.py:22]: The camera only outputs MJPEG and YUV2. We can compress it however we like with ffmpeg.

File: oresat_dxwifi/camera/camera.py [hunk 15]
```diff
@@ -1,115 +1,59 @@
...
-        video_x,
-        video_y,
-        fps,
-        spv,
-        br,
-        duration,
-        image_output,
-        video_output,
-        frames_del,
+        spv: int,
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py:16]: Let's call this `seconds_per_video` and call `br -> bit_rate`, since they're not common acronyms like `fps`. Not a big deal since their meanings are documented, but makes calls clearer.

File: oresat_dxwifi/camera/camera.py [hunk 16]
```diff
@@ -119,116 +63,144 @@ def __init__(
...
-            subprocess.call(capture_command.split())
 
-            # Call ffmpeg (non-blocking)
-            print("Starting encoding of video {} of {}.".format(i + 1, num_loops))
-            procs.append(subprocess.Popen(ffmpeg_command.split()))
+        try:
+            self._fresh_dir(self.image_output_directory)
+            self._fresh_dir(self.video_output_directory)
+        except Exception as directory_error:
+            raise CameraInterfaceError(directory_error)
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py:114]: I think it's better to simply raise the original exception in this case, as well as for the `capture_error` case farther down. Otherwise, we're hiding the original exception type; we're instantiating a new `CameraInterfaceError` using the original's `__str__` but losing the type info.

These aren't errors that we're finding or generating within the camera interface, they're errors that are bubbling up from elsewhere when the camera interface is trying to do its work.

Thoughts?

File: oresat_dxwifi/camera/camera.py [hunk 17]
```diff
@@ -119,116 +63,144 @@ def __init__(
...
-
-    def calculate_number_of_loops(self):
-        return self.total_duration // self.seconds_per_video
+            logger.info("Created new directory: {}".format(path))
 
-    def delete_frames(self):
-        # If requested, delete frames
+    def _delete_frames(self) -> None:
+        """If frame delete variable is true, deletes frames in image output directory"""
         if self.are_frames_deleted:
```
Comment (nrwahl2) [inline oresat_dxwifi/camera/camera.py:93]: Let's move the `if self.are_frames_deleted` to the caller, and let `_delete_frames()` just delete the frames

File: oresat_dxwifi/data/oresat_dxwifi.eds [hunk 18]
```diff
@@ -4191,6 +4192,13 @@ ParameterName=Temperature
 ObjectType=0x07
 DataType=0x0008
 AccessType=ro
+DefaultValue=0
+
+[6002]
+ParameterName=New Variable
+ObjectType=0x07
+DataType=0x0007
+AccessType=rw
```
Comment (nrwahl2) [inline oresat_dxwifi/data/oresat_dxwifi.eds]: `DefaultValue` and `PDOMapping` are missing here. compared to the dcf file. Do they belong here?

File: oresat_dxwifi/services/camera.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+class State(IntEnum):
+    OFF = 0
+    BOOT = 1
+    STANDBY = 2
+    FILMING = 3
+    ERROR = 0xFF
+
+
+STATE_TRANSISTIONS = {
```
Comment (nrwahl2) [inline oresat_dxwifi/services/camera.py]: `s/TRANSISTIONS/TRANSITIONS/`

File: oresat_dxwifi/services/camera.py [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+    State.OFF: [State.BOOT],
```
Comment (nrwahl2) [inline oresat_dxwifi/services/camera.py]: Is a transition back to `OFF` needed (perhaps from `ERROR`)? Not entirely sure how the state machine works

File: oresat_dxwifi/services/camera.py [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+
+
+class CameraService(Service):
+    def __init__(self):
+        super().__init__()
+        self.state = State.BOOT
+
+        self.spv = 3  # Should evenly divide into duration
+        self.duration = 6
+        self.IMAGE_OUPUT_DIRECTORY = "/oresat-live-output/frames"  # Make sure directory actually exists
```
Comment (nrwahl2) [inline oresat_dxwifi/services/camera.py]: Don't we create it if it doesn't?

File: oresat_dxwifi/services/camera.py [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+    State.ERROR: [],
+}
+
+
+class CameraService(Service):
+    def __init__(self):
+        super().__init__()
+        self.state = State.BOOT
+
+        self.spv = 3  # Should evenly divide into duration
```
Comment (nrwahl2) [inline oresat_dxwifi/services/camera.py]: A lot of these parameters are redundant -- we're setting members of `CameraService` but never using them, except to instantiate the `self.camera` member.

We could just pass these directly as arguments to the `CameraInterface` constructor.

File: oresat_dxwifi/services/camera.py [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+    OFF = 0
+    BOOT = 1
+    STANDBY = 2
+    FILMING = 3
+    ERROR = 0xFF
+
+
+STATE_TRANSISTIONS = {
+    State.OFF: [State.BOOT],
+    State.BOOT: [State.STANDBY],
```
Comment (nrwahl2) [inline oresat_dxwifi/services/camera.py]: Does `State.BOOT` really do anything for us? I see that `CameraService.on_start()` sets state to `State.STANDBY`. Wondering if both `OFF` and `BOOT` are necessary, or if we can use just one.

File: oresat_dxwifi/data/oresat_dxwifi.eds [hunk 24]
```diff
@@ -4191,6 +4192,13 @@ ParameterName=Temperature
 ObjectType=0x07
 DataType=0x0008
 AccessType=ro
+DefaultValue=0
+
+[6002]
+ParameterName=New Variable
```
Comment (nrwahl2) [inline oresat_dxwifi/data/oresat_dxwifi.eds]: ParameterName=Oresat Live Camera Resource Status

File: oresat_dxwifi/services/camera.py [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+            return self.state.value
+
+    def on_state_write(self, index: int, subindex: int, data):
+        if index != self.state_index:
+            return
+
+        try:
+            new_state = State(data)
+        except ValueError:
+            logger.error(f'Not a valid state: {data}')
```
Comment (nrwahl2) [inline oresat_dxwifi/services/camera.py]: We probably want to return here

**Review by nrwahl2 — COMMENTED** [review]
Progress

File: oresat_dxwifi/data/oresat_dxwifi.dcf [hunk 26]
```diff
@@ -2,7 +2,7 @@
 FileName=oresat_dxwifi.dcf
 FileVersion=1
 FileRevision=1
-LastEDS=oresat_dxwifi.eds
+LastEDS=
```
Comment (nrwahl2) [inline oresat_dxwifi/data/oresat_dxwifi.dcf]: The commit message doesn't make it clear why we're changing the `LastEDS`, `NodeID`, and `NodeName` fields -- or how they're related to the REST API.

File: oresat_dxwifi/data/oresat_dxwifi.dcf [hunk 27]
```diff
@@ -4204,8 +4204,18 @@ DataType=0x0008
...
+;Status. read for current state, write to change state
+;
+;States
+;
+;- 0x00: OFF
+;- 0x01: BOOT
+;- 0x02: STANDBY
+;- 0x03: FILM
+;- 0x04: TRANSMISSION
+; - 0xFF: ERROR
```
Comment (nrwahl2) [inline oresat_dxwifi/data/oresat_dxwifi.dcf]: drop the leading space here and in eds

File: oresat_dxwifi/data/oresat_dxwifi.eds [hunk 28]
```diff
@@ -4194,11 +4194,22 @@ DataType=0x0008
...
+;- 0x02: STANDBY
+;- 0x03: FILM
+;- 0x04: TRANSMISSION
+; - 0xFF: ERROR
 [6002]
-ParameterName=New Variable
+ParameterName=Oresat Live Service Status
 ObjectType=0x07
-DataType=0x0007
+DataType=0x0005
```
Comment (nrwahl2) [inline oresat_dxwifi/data/oresat_dxwifi.eds]: Why change `DataType` from 0x0007 to 0x0005?

File: oresat_dxwifi/services/oresat_live.py [hunk 29]
```diff
@@ -10,19 +10,21 @@ class State(IntEnum):
     BOOT = 1
     STANDBY = 2
     FILMING = 3
+    TRANSMISSION = 4
```
Comment (nrwahl2) [inline oresat_dxwifi/services/oresat_live.py:19]: If we're adding a state, that should happen in a separate commit. This commit is supposed to be for adding REST API integration.

File: oresat_dxwifi/services/oresat_live.py [hunk 30]
```diff
@@ -70,24 +72,24 @@ def capture(self):
...
+        if index != self.STATE_INDEX:
             return
 
         try:
             new_state = State(data)
         except ValueError:
             logger.error(f'Not a valid state: {data}')
 
-        if new_state == self.state or new_state in STATE_TRANSISTIONS(self.state):
+        if new_state == self.state or new_state in STATE_TRANSISTIONS[self.state]:
```
Comment (nrwahl2) [inline oresat_dxwifi/services/oresat_live.py]: "STATE_TRANSISTIONS" is a typo

File: oresat_dxwifi/webpage/oresat_live.html [hunk 31]
```diff
@@ -1,36 +0,0 @@
-<!DOCTYPE html>
```
Comment (nrwahl2) [inline oresat_dxwifi/webpage/oresat_live.html]: Why did we create this page in one commit just to remove it in a later commit of the same PR?

File: oresat_dxwifi/services/oresat_live.py [hunk 32]
```diff
@@ -56,12 +56,12 @@ def __init__(self):
         )
 
     def on_start(self):
-        self.STATE_INDEX = 0x6002
+        self.STATE_INDEX = "live_status"
```
Comment (nrwahl2) [inline oresat_dxwifi/services/oresat_live.py]: Does the `STATE_INDEX` in the html file's `<script>` section need to be updated too?

File: README.md [hunk 33]
```diff
@@ -22,6 +22,5 @@ video broadcast from space.
 - Make a virtual CAN bus
   - `$ sudo ip link add dev vcan0 type vcan`
   - `$ sudo ip link set vcan0 up`
-- Run `$ python3 -m oresat_dxwifi`
-  - Can mock the SkyTraq by adding the `-m` flag
+- Run `$ python -m oresat_dxwifi`
```
Comment (nrwahl2) [inline README.md:28]: Merging master in shouldn't be a separate commit; you should rebase your whole branch on the current master, resolve any merge conflicts, and push again. That doesn't affect the code though, so we can deal with that any time before we merge this PR.

File: oresat_dxwifi/templates/oresat_live.html [hunk 34]
```diff
@@ -0,0 +1,73 @@
...
+            <button onclick="state_write(State.OFF)">OFF</button>
+            <button onclick="state_write(State.BOOT)">BOOT</button>
+            <button onclick="state_write(State.STANDBY)">STANDBY</button>
+            <button onclick="state_write(State.FILMING)">FILMING</button>
+            <button onclick="state_write(State.TRANSMISSION)">TRANSMISSION</button>
+            <button onclick="state_write(State.ERROR)">ERROR</button>
+        </div>
+    </body>
+    <script>
+        const STATE_INDEX = 0x6002
```
Comment (nrwahl2) [inline oresat_dxwifi/templates/oresat_live.html]: Make this a string. See the function documentation for `readValue` and `writeValue`: https://github.com/oresat/oresat-olaf/blob/master/olaf/_internals/rest_api/templates/_base.html

File: oresat_dxwifi/templates/oresat_live.html [hunk 35]
```diff
@@ -0,0 +1,73 @@
...
+            OFF: 0,
+            BOOT: 1,
+            STANDBY: 2,
+            FILMING: 3,
+            TRANSMISSION: 4,
+            ERROR: 0xFF
+        }
+
+        async function state_read(){
+            const obj = await readValue('0x6002', null);
```
Comment (nrwahl2) [inline oresat_dxwifi/templates/oresat_live.html]: Use `STATE_INDEX`

File: oresat_dxwifi/data/oresat_dxwifi.eds (hunk 28 above)
Comment (bpranaw) [reply oresat_dxwifi/data/oresat_dxwifi.eds]: The .eds is deleted so this doesn't matter.

File: oresat_dxwifi/data/oresat_dxwifi.dcf (hunk 27 above)
Comment (bpranaw) [reply oresat_dxwifi/data/oresat_dxwifi.dcf]: The .dcf is being deleted so this doesn't matter.

File: oresat_dxwifi/data/oresat_dxwifi.dcf (hunk 26 above)
Comment (bpranaw) [reply oresat_dxwifi/data/oresat_dxwifi.dcf]: The .dcf is being deleted so this doesn't matter.

File: oresat_dxwifi/services/oresat_live.py (hunk 32 above)
Comment (bpranaw) [reply oresat_dxwifi/services/oresat_live.py]: Yeah. Done.

File: oresat_dxwifi/webpage/oresat_live.html (hunk 31 above)
Comment (bpranaw) [reply oresat_dxwifi/webpage/oresat_live.html]: Mistake

File: README.md (hunk 33 above)
Comment (bpranaw) [reply README.md:28]: Things got complicated because I had already committed my changes locally but hadn't synced with GitHub yet. I started the merging on GitHub, but ended up having to do it via the command line locally anyways. I pulled from the news master and merged into my local camera branch, resolved the issues and pushed it. I believe I got the option to do it via commit or something else. I didn't really see the difference since the merge issues were resolved and everything seems to be in order in both branches on GitHub.

File: oresat_dxwifi/templates/oresat_live.html [hunk 36]
```diff
@@ -26,7 +26,7 @@ <h2>Write</h2>
         </div>
     </body>
     <script>
-        const STATE_INDEX = "live_status"
+        const STATE_INDEX = "status"
```
Comment (nrwahl2) [inline oresat_dxwifi/templates/oresat_live.html:29]: If we change it here, we need to change it in `oresat_dxwifi/services/oresat_live.py`.

Comment (bpranaw) [reply oresat_dxwifi/templates/oresat_live.html:29]: Should already be changed

**Comment by bpranaw** [comment]
Going to be adding docstrings and notes into the code shortly.

File: oresat_dxwifi/templates/oresat_live.html (hunk 36 above)
Comment (nrwahl2) [reply oresat_dxwifi/templates/oresat_live.html:29]: yeah must've been a later commit that I haven't gotten to yet

File: oresat_dxwifi/transmission/defaults.py [hunk 37]
```diff
@@ -0,0 +1,111 @@
...
+    FROM_DS = True
+    MORE_FRAG = False
+    RETRY = False
+    POWER_MANAGEMENT = False
+    MORE_DATA = True
+    WEP = False
+    ORDER = False
+
+
+class TX():
```
Comment (nrwahl2) [inline oresat_dxwifi/transmission/defaults.py:80]: no parens on the classes that don't inherit

File: oresat_dxwifi/transmission/defaults.py [hunk 38]
```diff
@@ -0,0 +1,111 @@
+from enum import IntEnum
+
+
+# These are the default values from oresat-libdxwifi's tx.c/cli.h/transmitter.h
+# Not all of them are being used in transmission.py, but they'll be necessary to know
+# if oresat-libdxwifi's python bindins are to be integrated at a deeper level.
```
Comment (nrwahl2) [inline oresat_dxwifi/transmission/defaults.py:6]: s/bindins/bindings/

File: .gitignore [hunk 39]
```diff
@@ -145,4 +145,7 @@ dmypy.json
 *oresat_dxwifi/camera/lib/*
 
 # Camera Code: Output from camera
-oresat_dxwifi/camera/data/*
\ No newline at end of file
+oresat_dxwifi/camera/data/*
+
+# Transmission: Libdxwifi Python Bindings
+!oresat_dxwifi/transmission/tx_module.so
```
Comment (nrwahl2) [inline .gitignore]: We generally shouldn't include something that we build. Why are we tracking a compiled library as part of the git repo?

File: README.md [hunk 40]
```diff
@@ -17,10 +17,17 @@ video broadcast from space.
       - `$ cd build && cmake ..`
       - `$ make`
       - Back to root: cd `$ cd ../../../`
+      - Make output directories: `$ sudo mkdir /oresat-live-output/frames /oresat-live-output/videos`
 
 - Install dependenies `$ pip install -r requirements.txt`
```
Comment (nrwahl2) [inline README.md]: s/dependenies/dependencies/

File: oresat_dxwifi/services/oresat_live.py [hunk 41]
```diff
@@ -75,11 +78,32 @@ def capture(self):
...
+    def transmit(self):
+        self.state = State.TRANSMISSION
+
+        files = listdir(self.VIDEO_OUTPUT_DIRECTORY)
+
+        # Ideally, we could just pass in the output directory path to the Transmitter
+        # but in practice multi-file transmission to multi-file receiving has been inconsistent
+
+        for x in files:
+            x = self.VIDEO_OUTPUT_DIRECTORY + "/" + x
```
Comment (nrwahl2) [inline oresat_dxwifi/services/oresat_live.py]: `os.path.join`

File: oresat_dxwifi/services/oresat_live.py [hunk 42]
```diff
@@ -75,11 +78,32 @@ def capture(self):
...
+            try:
+                tx = Transmitter(x)
+                p = Process(target=tx.transmit)
```
Comment (nrwahl2) [inline oresat_dxwifi/services/oresat_live.py:143]: Include a comment about why you're using multiprocessing here; it's not obvious.

File: .gitignore (hunk 39 above)
Comment (bpranaw) [reply .gitignore]: What should we do instead?

Comment (nrwahl2) [reply .gitignore]: have the user build the .so file when they build everything else

Comment (nrwahl2) [reply .gitignore]:
```
debian@cs-capstone-bbb:~/src/oresat-libdxwifi$ cd build && cmake .. && make && cd ..
...
-- Found pybind11: /usr/include (found version "2.6.2" )
...
[ 81%] Built target tx
[ 83%] Building CXX object dxwifi/tx/CMakeFiles/tx_module.dir/tx_wrapper.cpp.o
[ 85%] Linking CXX shared module ../../../bin/Debug/tx_module.so
...
[100%] Built target decode
```

Comment (bpranaw) [reply .gitignore]: Wouldn't the defeat the purpose of the bindings? If we need to install libdxwifi anyways, we could just subprocess it since we have to anyways. Is there a way to host the library somewhere and include it as a dependency or something? Alternatively, we could just do what Ryan suggested and move libdxwifi here. I don't really have any preferences, I genuinely thought this is why we even wanted them in the first place, for convenient integration/easier build.

File: README.md (hunk 40 above)
Comment (bpranaw) [reply README.md]: ``
```suggestion
- Install dependencies `$ pip install -r requirements.txt`
```

File: oresat_dxwifi/camera/camera.py (hunk 1 above)
Comment (bpranaw) [reply oresat_dxwifi/camera/camera.py]: Turned into a static method and underlined.

File: oresat_dxwifi/camera/camera.py (hunk 2 above)
Comment (bpranaw) [reply oresat_dxwifi/camera/camera.py]: Outdated now, this part has been removed.

File: .gitignore (hunk 39 above)
Comment (nrwahl2) [reply .gitignore]: I didn't say anything about installation; I'm talking about build. When you build libdxwifi currently, it creates the executable binaries (e.g., `tx`, `rx`) and also the `tx_module.so` file. Then it's your choice what to do with those.

It's not a good idea to include in a git repo a blob that's already compiled. It should be compiled on the target machine as needed.

Ideally we should have a libdxwifi build option that lets you choose whether to create the `tx_module.so`, but currently we do not.

File: oresat_dxwifi/camera/camera.py (hunk 5 above)
Comment (bpranaw) [reply oresat_dxwifi/camera/camera.py]: Done.

File: .gitignore (hunk 39 above)
Comment (nrwahl2) [reply .gitignore]: Wait a second -- this isn't libdxwifi, this is oresat-dxwifi-software. Hmm, let me think.

File: oresat_dxwifi/camera/camera.py (hunk 10 above)
Comment (bpranaw) [reply oresat_dxwifi/camera/camera.py]: All the resolution stuff uses some variant of this now.

File: oresat_dxwifi/camera/camera.py (hunk 16 above)
Comment (bpranaw) [reply oresat_dxwifi/camera/camera.py:114]: Valid. At the time I just didn't want to rework fresh_dir and get straight to the camera system working. The only way it fails in theory I think is the directories don't exist.  I think it's low priority for now since the message goes through.

File: oresat_dxwifi/data/oresat_dxwifi.eds (hunk 18 above)
Comment (bpranaw) [reply oresat_dxwifi/data/oresat_dxwifi.eds]: File no longer necessary.

File: oresat_dxwifi/services/camera.py (hunk 19 above)
Comment (bpranaw) [reply oresat_dxwifi/services/camera.py]: Typo fixed.

File: oresat_dxwifi/camera/camera.py (hunk 16 above)
Comment (nrwahl2) [reply oresat_dxwifi/camera/camera.py:114]: Just add TODO comments

File: oresat_dxwifi/services/camera.py (hunk 21 above)
Comment (bpranaw) [reply oresat_dxwifi/services/camera.py]: No we just don't capture. The way the capture stuff is set up down the line just expects it I guess. Still something to consider changing since we set the video configurations here now.

File: oresat_dxwifi/services/camera.py (hunk 22 above)
Comment (bpranaw) [reply oresat_dxwifi/services/camera.py]: A lot of this stuff still needs cleanup because it tried to maintain the integrity of oresat-live-software/camera cmd functionality.

File: oresat_dxwifi/services/camera.py (hunk 21 above)
Comment (nrwahl2) [reply oresat_dxwifi/services/camera.py]: What about in `create_videos()` where it calls `self._fresh_dir(self.image_output_directory)`?

File: oresat_dxwifi/services/camera.py (hunk 23 above)
Comment (bpranaw) [reply oresat_dxwifi/services/camera.py]: I was following how star-tracker implemented theirs. I think it is fine to leave for now because it leaves a distinction between the states in case we need to add more to what they do in the future.

File: oresat_dxwifi/data/oresat_dxwifi.eds (hunk 24 above)
Comment (bpranaw) [reply oresat_dxwifi/data/oresat_dxwifi.eds]: File is no longer necessary.

File: oresat_dxwifi/services/oresat_live.py (hunk 29 above)
Comment (bpranaw) [reply oresat_dxwifi/services/oresat_live.py:19]: I believe this was around when we need to integrate Olaf v3 so I mixed up some stuff. My bad.

File: oresat_dxwifi/services/camera.py (hunk 21 above)
Comment (bpranaw) [reply oresat_dxwifi/services/camera.py]: Oh I misunderstood what you were referring to. That doesn't create the directories, it just clears them. It needs them to exist. Its internal logic uses a receiving an error to determine whether to remove a directory or file. I think that's why I didn't try to rework it.

Comment (bpranaw) [reply oresat_dxwifi/services/camera.py]: It's a remaining component from oresat-live-software/camera

Comment (nrwahl2) [reply oresat_dxwifi/services/camera.py]:
```
    def create_videos(self) -> None:
            ...
            self._fresh_dir(self.image_output_directory)

    def _fresh_dir(path) -> None:
        if os.path.exists(path):
            ...
        else:
            # Make
            os.mkdir(path)
```

File: oresat_dxwifi/services/camera.py (hunk 25 above)
Comment (bpranaw) [reply oresat_dxwifi/services/camera.py]: You're right. I'm adding it in the next commit.

File: oresat_dxwifi/services/oresat_live.py (hunk 30 above)
Comment (bpranaw) [reply oresat_dxwifi/services/oresat_live.py]: Typo fixed.

File: README.md [hunk 43]
```diff
@@ -5,9 +5,29 @@ video broadcast from space.
 
 ## Quickstart
 
-- Install dependenies `$ pip install -r requirements.txt`
+- Install Python and pip
+  - `$ sudo apt update`
+  - `$ sudo apt install python3 python3-pip`
+
+  ### Build Camera Code
+  - Install dependenies for camera folder. Check [camera/README.md](./oresat_dxwifi/camera/README.md) for more information
```
Comment (nrwahl2) [inline README.md:13]: s/dependenies/dependencies/

File: oresat_dxwifi/services/oresat_live.py [hunk 44]
```diff
@@ -86,7 +97,8 @@ def capture(self):
             logger.error("Something went wrong with camera capture...")
             logger.error(error)
 
-    def transmit(self):
+    def transmit(self) -> None:
+        "Transmits all the videos in the video output directory"
```
Comment (nrwahl2) [inline oresat_dxwifi/services/oresat_live.py:124]: triple-quote

File: oresat_dxwifi/services/oresat_live.py [hunk 45]
```diff
@@ -107,9 +119,18 @@ def transmit(self):
         self.state = State.STANDBY
 
     def on_state_read(self) -> State:
+        """Returns the current state value. Function is will be called on an SDO read of status
```
Comment (nrwahl2) [inline oresat_dxwifi/services/oresat_live.py:152]: s/is will be called/is called/

File: README.md [hunk 46]
```diff
@@ -20,10 +20,6 @@ video broadcast from space.
       - Make output directories: `$ sudo mkdir /oresat-live-output/frames /oresat-live-output/videos`
 
 - Install dependencies `$ pip install -r requirements.txt`
-  - (Temporary) Install forked oresat-configs
```
Comment (nrwahl2) [inline README.md]: Do we still need it in requirements.txt?

**Review by nrwahl2 — COMMENTED** [review]
thanks, working on pulling these into my branch but I haven't yet reviewed the "Fixed state" and "Fixed mistake" commits
